File: crates/kremory-napi/examples/offline-remember-recall.mjs
// **Start here.** Save something, get it back — offline, in about thirty seconds.
//
// Node mirror of `crates/kremory/examples/offline_remember_recall.rs`, scenario
// slug `offline-remember-recall`. The two are kept in lockstep by
// `scripts/check-sdk-scenarios.sh`, which runs BOTH and asserts they agree — so
// the docs-site language tabs cannot show a Node snippet that no longer works.
//
// No Ollama. No API keys. No environment variables. No network.
//
// ⚠️ Opening offline needs BOTH a custom embedder AND a custom extractor.
// Supplying only `withEmbedder` takes a different branch that hard-requires an
// env LLM (`crates/kremory-napi/src/lib.rs`, the `resolve_env_llm()` call after
// the `extractor_handle` block) — which is why `09-byom-embedder.mjs` cannot run
// without one. Supplying an extractor as well selects the no-LLM path.
//
// ⚠️ `@kgentic-ai/kremory-node` is UNPUBLISHED (v0.4.1) — this imports the
// LOCALLY BUILT native module, not a published npm package.
//
// REQUIRES: native module built (`pnpm build:debug` in crates/kremory-napi/).
//
// Run: `node examples/offline-remember-recall.mjs`

import assert from 'node:assert/strict';
import os from 'node:os';
import path from 'node:path';
import fs from 'node:fs';
import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);
const { Memory } = require('../index.js');

const DEMO_DIM = 16;
const NAMESPACE = 'demo';

/**
 * A deterministic stand-in so this example needs no embedding service.
 * **Not for production** — it hashes bytes, it does not understand meaning.
 * A real consumer calls their embedding backend here (OpenAI, Ollama
 * `nomic-embed-text`, a local GGUF, sentence-transformers over HTTP).
 *
 * Byte-for-byte the same arithmetic as `DemoEmbedder` in the Rust example, so
 * both SDKs produce identical vectors for identical text.
 */
function demoEmbedder(dim) {
  // ⚠️ TWO arguments. The embedder is invoked through a napi-rs
  // ThreadsafeFunction with the error-first calling convention, exactly like
  // `opts.extractor.extract` — `err` is always null and the TEXT IS THE SECOND
  // ARGUMENT. A one-argument `async (text) => …` silently binds `text` to null,
  // and a null-guard that maps it to '' then returns an ALL-ZERO vector of the
  // right length. It passes the bridge's dimension check, stores cleanly, and
  // has zero cosine similarity with every query — so entities embed to nothing
  // and become permanently unrecallable, with no error anywhere.
  return async function embed(_err, text) {
    const safeText = text == null ? '' : String(text);
    const vec = new Array(dim).fill(0);
    for (let i = 0; i < safeText.length; i++) {
      vec[i % dim] += safeText.charCodeAt(i) / 255.0;
    }
    const norm = Math.sqrt(vec.reduce((s, v) => s + v * v, 0)) || 1e-6;
    return vec.map((v) => v / norm);
  };
}

/**
 * Never invoked — the facts below are supplied directly and `skipExtraction`
 * turns Phase-2 off. It exists because opening without an LLM requires saying
 * HOW facts would be extracted, exactly as the Rust example's `NoExtraction`
 * does.
 *
 * NB `extract` takes napi-rs's error-first arguments: `err` is always null and
 * the text is the SECOND argument.
 */
const noExtraction = {
  name: 'no-extraction',
  extract: async (_err, _text) => ({ entities: [], facts: [] }),
};

const dbPath = path.join(os.tmpdir(), `kremory-example-offline-${Date.now()}.db`);

const mem = await Memory.open(dbPath, {
  embeddingDim: DEMO_DIM,
  defaultNamespace: NAMESPACE,
  withEmbedder: demoEmbedder(DEMO_DIM),
  extractor: noExtraction,
});

try {
  // Save two things we know about a user.
  await mem.remember({
    content: 'Notes about Jim.',
    namespace: NAMESPACE,
    structuredFacts: [
      { subject: 'jim', predicate: 'writes', object: 'Rust' },
      { subject: 'jim', predicate: 'prefers', object: 'concise replies' },
    ],
    skipExtraction: true,
  });

  // Get them back. `recall()` gives you the structured facts;
  // `recallAsPromptText()` gives you a rendered string ready to drop into a
  // prompt instead.
  const contexts = await mem.recall('what do we know about jim', { namespace: NAMESPACE });
  const facts = contexts.flatMap((c) => c.facts);

  for (const f of facts) {
    console.log(`${f.subject} ${f.predicate} ${f.object}`);
  }

  assert.ok(
    facts.length > 0,
    'expected to recall the facts we just stored. If this fails with KREMORY_* ' +
      'environment variables set, unset them — they override recall scoring.',
  );

  const rendered = await mem.recallAsPromptText('what do we know about jim', null, {
    namespace: NAMESPACE,
  });
  console.log(`\n--- prompt-ready ---\n${rendered}`);
} finally {
  await mem.close();
  try { fs.unlinkSync(dbPath); } catch { /* best-effort cleanup */ }
}

// A custom `withEmbedder` callback (this example uses one, per the module
// doc above) leaves a napi-rs ThreadsafeFunction handle that keeps the
// event loop alive past process completion — the same known bridge quirk
// `09-byom-embedder.mjs` already force-exits for. Discovered here via
// `scripts/check-sdk-scenarios.sh` (the process printed correct output,
// including the closing "prompt-ready" block, then never exited on its
// own). Force exit once everything above has genuinely completed.
process.exit(0);

File: crates/kremory/examples/offline_remember_recall.rs
//! **Start here.** Save something, get it back — offline, in about thirty seconds.
//!
//! ```text
//! cargo run --example offline_remember_recall
//! ```
//!
//! No Ollama. No API keys. No environment variables. No network. Nothing to install
//! beyond the crate itself, and nothing left on disk when it exits.
//!
//! This is the smallest useful kremory program. Once it makes sense, read
//! `remembers_across_sessions.rs`, which shows the part no other memory store does:
//! answering "where does Alice live?" and "where did she live in March?" correctly
//! at the same time.
//!
//! ## Why an "extractor" appears in a program that extracts nothing
//!
//! kremory's normal job is to read prose and work out the facts itself, which needs a
//! language model. Here we supply the facts directly and call `.skip_extraction()`,
//! so no model is needed — but the builder still requires you to say HOW facts would
//! be extracted. Hence the do-nothing extractor below. It is never called.

use std::future::Future;
use std::sync::Arc;

use kremory::core::intelligence::{EntityExtractor, ExtractionContext, ExtractionResult};
use kremory::{CoreResult, EmbeddingProvider, Memory, Namespace, StructuredFact};

const DEMO_DIM: usize = 16;

/// A deterministic stand-in so this example needs no embedding service.
/// **Not for production** — it hashes bytes, it does not understand meaning.
/// A real consumer calls their embedding backend here (OpenAI, Ollama
/// `nomic-embed-text`, a local GGUF, sentence-transformers over HTTP).
struct DemoEmbedder {
    dim: usize,
}

impl EmbeddingProvider for DemoEmbedder {
    fn embed<'a>(
        &'a self,
        text: &'a str,
    ) -> impl Future<Output = CoreResult<Vec<f32>>> + Send + 'a {
        let dim = self.dim;
        async move {
            let mut v = vec![0f32; dim];
            for (i, b) in text.bytes().enumerate() {
                v[i % dim] += f32::from(b) / 255.0;
            }
            let norm = v.iter().map(|x| x * x).sum::<f32>().sqrt().max(1e-6);
            for x in &mut v {
                *x /= norm;
            }
            Ok(v)
        }
    }
}

/// Never invoked — see the module doc. Wiring a real one is the "bring your own
/// extractor" path: return the entities and facts you found, and kremory stores them.
struct NoExtraction;

impl EntityExtractor for NoExtraction {
    fn name(&self) -> &'static str {
        "no-extraction"
    }

    // `async fn` rather than `-> impl Future`: the trait is AFIT-shaped, so this is
    // the simpler spelling and the one clippy's `manual_async_fn` asks for.
    async fn extract(
        &self,
        _text: &str,
        _ctx: &ExtractionContext<'_>,
    ) -> CoreResult<ExtractionResult> {
        Ok(ExtractionResult {
            entities: Vec::new(),
            facts: Vec::new(),
        })
    }
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let dir = tempfile::tempdir()?;
    let ns = Namespace::new("demo");

    let mem = Memory::open(dir.path().join("memory.db"))
        .embedding_dim(DEMO_DIM)
        .default_namespace(ns.clone())
        .with_embedder(Arc::new(DemoEmbedder { dim: DEMO_DIM }))
        .with_extractor(Arc::new(NoExtraction))
        .await?;

    // Save two things we know about a user.
    mem.remember("Notes about Ada.")
        .in_namespace(ns.clone())
        .with_facts(vec![
            StructuredFact {
                subject: "jim".into(),
                predicate: "writes".into(),
                object: "Rust".into(),
                valid_from: None, // None = "true as of now"
                valid_to: None,   // None = "still true"
                memory_type: None,
            },
            StructuredFact {
                subject: "jim".into(),
                predicate: "prefers".into(),
                object: "concise replies".into(),
                valid_from: None,
                valid_to: None,
                memory_type: None,
            },
        ])
        .skip_extraction()
        .await?;

    // Get them back. `.raw()` gives you the structured facts; a bare `.recall(..)`
    // gives you a rendered String ready to drop into a prompt instead.
    let facts: Vec<_> = mem
        .recall("what do we know about jim")
        .in_namespace(ns.clone())
        .raw()
        .await?
        .into_iter()
        .flat_map(|c| c.facts)
        .collect();

    for f in &facts {
        println!("{} {} {}", f.subject, f.predicate, f.object);
    }

    assert!(
        !facts.is_empty(),
        "expected to recall the facts we just stored. If this fails with KREMORY_* \
         environment variables set, unset them — they override recall scoring."
    );

    // And the prompt-ready rendering, which is what most agents actually want.
    let rendered = mem
        .recall("what do we know about jim")
        .in_namespace(ns)
        .await?;
    println!("\n--- prompt-ready ---\n{rendered}");

    mem.close().await?;
    Ok(())
}
